wrappers/services.go

```diff
@@ -677,21 +703,58 @@ func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSn
 			return err
 		}
 	}
+	return nil
+}
 
-	if !preseeding {
-		if modifiedSystem {
-			if err = sysd.DaemonReload(); err != nil {
-				return err
-			}
-		}
-		if modifiedUser {
-			if err = userDaemonReload(); err != nil {
-				return err
-			}
+// EnsureSnapServices will ensure that the specified snap services' file states
+// are up to date with the specified options and infos. It will add new services
+// if those units don't already exist, but it does not delete existing service
+// units that are not present in the snap's Info structures.
+// There are two sets of options; there are global options which apply to the
+// entire transaction and to every snap service that is ensured, and options
+// which are per-snap service and specified in the map argument.
+// If any errors are encountered trying to update systemd units, then all
+// changes performed up to that point are rolled back, meaning newly written
+// units are deleted and modified units are attempted to be restored to their
+// previous state.
+// To observe which units were added or modified a
+// ObserveChangeCallback calllback can be provided. The callback is
+// invoked while processing the changes. Because of that it should not
+// produce immediate side-effects, as the changes are in effect only
+// if the function did not return an error.
+// This function is idempotent.
+func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSnapServicesOptions, observeChange ObserveChangeCallback, inter Interacter) (err error) {
+	if opts == nil {
+		opts = &EnsureSnapServicesOptions{}
+	}
+
+	context := &ensureSnapServicesContext{
+		snaps:         snaps,
+		observeChange: observeChange,
+		opts:          opts,
+		inter:         inter,
+		sysd:          systemd.New(systemd.SystemMode, inter),
+		modifiedUnits: make(map[string]*osutil.MemoryFileState),
+	}
+
+	defer func() {
+		if err == nil {
+			return
 		}
+		context.restore()
+	}()
+
+	quotaGroups, err := context.ensureSnapsSystemdServices()
+	if err != nil {
+		return err
 	}
 
-	return nil
+	err = context.ensureSnapSlices(quotaGroups)
```

Comment:
(nitpick^3) this could be a single line: `if err := context.ensureSnapSlices(quotaGroups); err != nil {`